crates/iota-core/src/authority/shared_object_congestion_tracker.rs

```diff
@@ -11,119 +11,294 @@ use iota_types::{
-            mode,
+        if self.assign_min_free_execution_slot {
+            // If `assign_min_free_execution_slot` is true, we assign the transaction start
+            // time based on the lowest free execution slot that can accommodates the
+            // transaction. We start the search from the full range of the slots
+            // available with no constraints from previous objects.
+            let initial_free_slot = ExecutionSlot::max_duration_non_scheduled_slot();
+            self.compute_min_free_execution_slot(shared_input_objects, tx_cost, initial_free_slot)
+                .unwrap_or(MAX_EXECUTION_TIME)
+        } else {
+            // If `assign_min_free_execution_slot` is false, we assign the transaction start
+            // time based on the maximum end time of the occupied execution slots for the
+            // transaction over all its shared objects.
+            shared_input_objects
+                .iter()
+                .map(|obj| {
+                    self.object_execution_slots
+                        .get(&obj.id)
+                        .expect("object should have been inserted at the start of this function.")
+                })
+                .map(|slots| max_object_free_slot_start_time(slots))
+                .max()
+                .expect("There must be at least one object in shared_input_objects.")
         }
     }
 
-    // Given a list of shared input objects, returns the starting cost of a
-    // transaction that operates on these objects.
-    //
-    // Starting cost is a proxy for the starting time of the transaction. It is
-    // determined by all the input shared objects' last write.
-    pub fn compute_tx_start_at_cost(&self, shared_input_objects: &[SharedInputObject]) -> u64 {
-        shared_input_objects
-            .iter()
-            .map(|obj| *self.object_execution_cost.get(&obj.id).unwrap_or(&0))
-            .max()
-            .expect("There must be at least one object in shared_input_objects.")
+    // A recursive function that tries to find the lowest free slot for a
+    // transaction. If a slot is found that fits the transaction, the function
+    // returns the slot. Otherwise, it returns None.
+    // lookup_interval is the range of the slot that the transaction can fit in
+    // given the objects that have been checked so far.
+    fn compute_min_free_execution_slot(
+        &self,
+        shared_input_objects: &[SharedInputObject],
+        tx_cost: u64,
+        lookup_interval: ExecutionSlot,
+    ) -> Option<ExecutionTime> {
+        // take the first object from the shared input objects.
+        let obj = shared_input_objects
+            .first()
+            .expect("shared_input_objects must not be empty.");
+        // set aside the remaining objects for the next recursive call.
+        let remaining_objects = if shared_input_objects.len() > 1 {
+            &shared_input_objects[1..]
+        } else {
+            &[]
+        };
```

Comment:
This can be nicely shortened:

```suggestion
        // Take the first object from the shared input objects, and
        // set aside the remaining objects for the next recursive call.
        let (first_object, remaining_objects) = shared_input_objects
            .split_first()
            .expect("shared_input_objects must not be empty.");
```